File: solutions/java/need-for-speed/1/src/main/java/NeedForSpeed.java
class NeedForSpeed {
  private int totalBattery = 100;
  private int totalDistance = 0;
  private final int batteryDrain;
  private final int speed;

  NeedForSpeed(int speed, int batteryDrain) {
    this.speed = speed;
    this.batteryDrain = batteryDrain;
  }

  public boolean batteryDrained() {
    return totalBattery < batteryDrain;
  }

  public int distanceDriven() {
    return totalDistance;
  }

  public void drive() {
    if (this.batteryDrained())
      return;
    totalBattery -= batteryDrain;
    totalDistance += speed;
  }

  public static NeedForSpeed nitro() {
    return new NeedForSpeed(50, 4);
  }
}

class RaceTrack {
  private int distance;

  RaceTrack(int distance) {
    this.distance = distance;
  }

  public boolean canFinishRace(NeedForSpeed car) {
    while (!car.batteryDrained())
      car.drive();
    return car.distanceDriven() >= this.distance;
  }
}

File: solutions/csharp/need-for-speed/1/NeedForSpeed.cs
class RemoteControlCar
{
  private int _batteryPercentage = 100;
  private int _distanceDriven = 0;
  private int _batteryDrain;
  private int _speed;

  public RemoteControlCar(int speed, int batteryDrain)
  {
    // Guard against infinite loops.
    _speed = speed > 0 ? speed : 1;
    _batteryDrain = batteryDrain > 0 ? batteryDrain : 1;
  }

  public bool BatteryDrained()
  {
    return _batteryPercentage < _batteryDrain;
  }

  public int DistanceDriven()
  {
    return _distanceDriven;
  }

  public void Drive()
  {
    if (!BatteryDrained())
    {
      _batteryPercentage -= _batteryDrain;
      _distanceDriven += _speed;
    }
  }

  public static RemoteControlCar Nitro()
  {
    return new RemoteControlCar(50, 4);
  }
}

class RaceTrack
{
  private int _distance;

  public RaceTrack(int distance)
  {
    _distance = distance;
  }

  public bool TryFinishTrack(RemoteControlCar car)
  {
    while (!car.BatteryDrained())
    {
      car.Drive();
    }
    return car.DistanceDriven() >= _distance;
  }
}
